sklearn_numba_dpex/common/kernels.py

```diff
@@ -397,7 +429,149 @@ def partial_sum_reduction(
                 local_values[zero_idx] + local_values[one_idx]
             )
 
-    return work_group_size, partial_sum_reduction
+    if return_only_kernel:
+        return partial_sum_reduction
+
+    partial_sum_reductiom_nofunc = _make_partial_sum_reduction_2d_axis1_kernels(
+        n_rows, work_group_size, None, dtype, device, return_only_kernel=True
+    )
+
+    return work_group_size, (
+        (partial_sum_reduction, partial_sum_reductiom_nofunc),
+        reduction_block_size,
+    )
+
+
+@lru_cache
+def _make_partial_sum_reduction_2d_axis0_kernels(
+    n_cols,
+    work_group_size,
+    sub_group_size,
+    fused_unary_func,
+    dtype,
+    device,
+    return_only_kernel=False,
+):
```

Comment:
There's already a docstring that sums up the approach in the top level function `make_sum_reduction_2d_kernel`:
```
    The strategy consists in performing local reductions in each work group using local
    memory where each work item combine two values, thus halving the number of values,
    and the number of active work items. At each iteration the work items are discarded
    in a bracket manner. The work items with the greatest ids are discarded first, and
    we rely on the fact that the remaining work items are adjacents to optimize the RW
    operations.

    Once the reduction is done in a work group the result is written in global memory,
    thus creating an intermediary result whose size is divided by
    `2 * work_group_size` if axis=1, or by `2 * work_group_size // sub_group_size` if\
    axis=0. This is repeated as many times as needed until only one value remains in
    global memory.
```

I think it's better there since it applies to implementations for both axis ?

For more in-depth explanations, since there are many technical details, I propose instead citing litterature. I haven't found a piece that really stands out as a better read than the others, but at least it converges on the same ideas:

```
    The algorithmic approach that is chosen to implement the underlying kernels takes
    inspiration from many sources that are available online, such as [1]_, [2]_, [3]_ .
    
    .. [1] Timcheck, S. W. (2017). Efficient Implementation of Reductions on GPU 
    Architectures.
    
    .. [2] https://dournac.org/info/gpu_sum_reduction
    
    .. [3] https://shreeraman-ak.medium.com/parallel-reduction-with-cuda-d0ae10c1ae2c
```

Reply:
My objective would be to make the reader of each individual kernel able to straightforwardly understand how the parallelism work without having to read any external reference.

It's good to have a high level description with reference in the top level API but it would be even better to also have short and concrete and specific inline comments in the code of the kernel to avoid any indirection when reading it.

Reply:
> My objective would be to make the reader of each individual kernel able to straightforwardly understand how the parallelism work without having to read any external reference.

I disagree on this objective. I think readers that arrive at this (low) level of code and need to maintain, debug, or improve, or formulate criticism for some reason, should pass some minimal level of required knowledge on gpu programming. We're already way above average level of commenting per line of code in this repository. I see the point for difficult kernels such as lloyd or topk and I happily provide more commenting there, since even with resources it requires some mental gymnastics to mirror the code and the papers, but this kernel sum is much simpler. 

So maybe you have other readers in mind ? such as students ? but if these implementations are going to serve as educational content it should be explicited as a target of the project. I don't have this in mind, and I think it would slow down development overall.

Maybe we can open a separate issue to discuss this and start contributing standards for the repo ?

Another thing we can discuss is how much time of review is considered OK per line of code in a PR.

Reply:
By the way @ogrisel if you want more perspective on the approach but don't have time to skim through the papers I'd happily provide you with live explanations, maybe we can just browse the code together.

Reply:
I've added exhaustive commenting so the PR can move on.

I'm :-1: for doing it again in the future for this amount of lines ~ took almost half a day and I think the reader still needs a pen and a paper to understand what is going on (in all the kernels the grid of work groups is mapped to the data is a way that a drawing makes it a lot more telling than words).